### PR TITLE
feature(embed): adds serve-icon page handler

### LINKED ISSUE #9582: How to handle embeds?
#9572 breaks our embed logic, mainly because our embed logic is flawed in that it relies on URLs to never change.

Strawman 1. Manually set icon/download URLs to use old handlers until we figure out a solution
Strawman 2. We can allow non-expiring URLs in file service (either by setting very high expiration time, or no expiration time; removing the last updated check for certain URLs)
Strawman 3. Use regex callbacks to replace URLs with dynamic equivalents

- [ ] Add ``serve-icon`` handler once #9655 is merged

### PR DESCRIPTION
**Pull request by hypeJunction**
Adds a handler to serve embedded icons. Adds elgg_get_embed_url() to
generate embed URLs

Fixes #9582

### CODE REVIEW

**Comment by hypeJunction** [comment]
At last the fix that started the icon epic. After this I can sleep knowing that 2.2 can be released.

File: engine/classes/Elgg/EntityIconService.php [hunk 1]
```diff
@@ -426,4 +437,66 @@ public function getSizes($entity_type = null, $entity_subtype = null, $type = 'i
...
+		list(, $guid, $size) = $m;
+
+		$entity = $this->entities->get($guid);
+		if (!$entity instanceof \ElggEntity) {
+			return $response->setStatusCode(404)->setContent('Item does not exist');
+		}
+
+		$thumbnail = $entity->getIcon($size);
+		if (!$thumbnail->exists()) {
+			return $response->setStatusCode(404)->setContent('Icon doest not exist');
```
Comment (mrclay) [inline engine/classes/Elgg/EntityIconService.php]: typo

Comment (hypeJunction) [inline engine/classes/Elgg/EntityIconService.php]: Off to bed. Will fix tomorrow with the other typos you shall find ;)

File: engine/classes/Elgg/EntityIconService.php [hunk 2]
```diff
@@ -426,4 +437,66 @@ public function getSizes($entity_type = null, $entity_subtype = null, $type = 'i
...
+		if (!empty($if_none_match)) {
+			// strip mod_deflate suffixes
+			$this->request->headers->set('if_none_match', str_replace('-gzip', '', $if_none_match));
+		}
+
+		$filenameonfilestore = $thumbnail->getFilenameOnFilestore();
+		$last_updated = filemtime($filenameonfilestore);
+		$etag = '"' . $last_updated . '"';
+		$response->setPublic()->setEtag($etag);
+		if ($response->isNotModified($this->request)) {
```
Comment (mrclay) [inline engine/classes/Elgg/EntityIconService.php]: In the 304 case, it's sending far off Expires headers. Probably should set +1 day here, too.

Comment (mrclay) [inline engine/classes/Elgg/EntityIconService.php]: Same with 404. if I get a 404 image, logging in doesn't help; I have to manually refresh.

Comment (mrclay) [inline engine/classes/Elgg/EntityIconService.php]: For 404 we don't want caching at all.

File: engine/classes/Elgg/EntityIconService.php [hunk 3]
```diff
@@ -426,4 +437,66 @@ public function getSizes($entity_type = null, $entity_subtype = null, $type = 'i
...
+		$if_none_match = $this->request->headers->get('if_none_match');
+		if (!empty($if_none_match)) {
+			// strip mod_deflate suffixes
+			$this->request->headers->set('if_none_match', str_replace('-gzip', '', $if_none_match));
+		}
+
+		$filenameonfilestore = $thumbnail->getFilenameOnFilestore();
+		$last_updated = filemtime($filenameonfilestore);
+		$etag = '"' . $last_updated . '"';
+		$response->setPublic()->setEtag($etag);
```
Comment (mrclay) [inline engine/classes/Elgg/EntityIconService.php]: setPrivate(). We don't want this shared on a proxy cache, etc.

File: engine/classes/Elgg/EntityIconService.php (hunk 2 above)
Comment (mrclay) [inline engine/classes/Elgg/EntityIconService.php]: Ideally we'd have test coverage for everything I've mentioned so far.

**Comment by mrclay** [comment]
Replaced by #9740

**Pull request closed without merging**